pkg/io/v2/providers/providers.go

```diff
@@ -0,0 +1,235 @@
+	googleCredentials, err := google.CredentialsFromJSON(ctx, credentials, storage.ScopeFullControl)
+	if err != nil {
+		return nil, fmt.Errorf("error getting Google credentials from JSON: %v", err)
+	}
+
+	client, err := gcp.NewHTTPClient(
+		gcp.DefaultTransport(),
+		gcp.TokenSource(googleCredentials.TokenSource))
+	if err != nil {
+		return nil, fmt.Errorf("error creating GCP Http Client: %v", err)
+	}
+
+	bkt, err := gcsblob.OpenBucket(ctx, client, bucketName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error opening GCS bucket: %v", err)
+	}
+	return bkt, nil
+}
+
+type s3Credentials struct {
+	Region           string `json:"region"`
+	Bucket           string `json:"bucket"`
+	Endpoint         string `json:"endpoint"`
+	Insecure         bool   `json:"insecure"`
+	S3ForcePathStyle bool   `json:"s3_force_path_style"`
+	AccessKey        string `json:"access_key"`
+	SecretKey        string `json:"secret_key"`
+}
+
+func getS3Bucket(ctx context.Context, creds []byte, bucketName string) (*blob.Bucket, error) {
+	s3Credentials := &s3Credentials{}
+	if err := json.Unmarshal(creds, s3Credentials); err != nil {
+		return nil, fmt.Errorf("error getting S3 credentials from JSON: %v", err)
+	}
+
+	staticCredentials := credentials.NewStaticCredentials(s3Credentials.AccessKey, s3Credentials.SecretKey, "")
+
+	sess, err := session.NewSession(&aws.Config{
+		Credentials:      staticCredentials,
+		Endpoint:         aws.String(s3Credentials.Endpoint),
+		DisableSSL:       aws.Bool(s3Credentials.Insecure),
+		S3ForcePathStyle: aws.Bool(s3Credentials.S3ForcePathStyle),
+		Region:           aws.String(s3Credentials.Region),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error creating S3 Session: %v", err)
+	}
+
+	bkt, err := s3blob.OpenBucket(ctx, sess, bucketName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error opening S3 bucket: %v", err)
+	}
+	return bkt, nil
+}
+
+func SignedURL(ctx context.Context, credentials []byte, storagePath string, opts *blob.SignedURLOptions) (string, error) {
+	storageProvider, bucketName, relativePath, err := ParseStoragePath(storagePath)
+	if err != nil {
+		return "", err
+	}
+
+	switch storageProvider {
+	case providerFile, providerS3:
+		bucket, err := GetBucket(ctx, credentials, storagePath)
+		if err != nil {
+			return "", err
+		}
+		return bucket.SignedURL(ctx, relativePath, opts)
+	case providerGS:
+		if len(credentials) == 0 {
+			artifactLink := &url.URL{
+				Scheme: httpsScheme,
+				Host:   "storage.googleapis.com",
+				Path:   path.Join(bucketName, relativePath),
+			}
+			return artifactLink.String(), nil
+		}
+		bucket, err := GetBucket(ctx, credentials, bucketName)
+		if err != nil {
+			return "", err
+		}
+		return bucket.SignedURL(ctx, relativePath, opts)
+	default:
+		return "", fmt.Errorf("unknown storageProvider: %s", storageProvider)
+	}
+}
+
+func ParseStoragePath(storagePath string) (storageProvider, bucket, relativePath string, err error) {
```

Comment:
This function is unused

Reply:
Nope, it's still used by providers.GetBucket  and opener.GetBucket